rio/src/RIO/Deque.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TypeFamilies #-}
+module RIO.Deque
+    ( -- * Types
+      Deque
+    , UDeque
+    , SDeque
+    , BDeque
+      -- * Operations
+    , newDeque
+    , popFrontDeque
+    , popBackDeque
+    , pushFrontDeque
+    , pushBackDeque
+    , foldlDeque
+    , foldrDeque
+    , dequeToList
+      -- * Inference helpers
+    , asUDeque
+    , asSDeque
+    , asBDeque
+    ) where
+
+import           RIO.Prelude.Reexports
+import           Control.Exception            (assert)
+import           Control.Monad                (liftM)
+import qualified Data.Vector.Generic.Mutable  as V
+import qualified Data.Vector.Mutable          as B
+import qualified Data.Vector.Storable.Mutable as S
+import qualified Data.Vector.Unboxed.Mutable  as U
+import           Data.Primitive.MutVar
+
+data DequeState v s a = DequeState
+    (v s a)
```

Comment:
Out of curiosity, why is the vector field not strict? I noticed it in the original code too, is it better for some reason to keep it lazy?

Reply:
It's an interesting question. Firstly, the primary performance benefit of making it strict (unpacking) wouldn't be possible, since it's a polymorphic field. The correctness argument of strict fields (avoiding bottom values and unevaluated values) would apply, and is a plus. There _may_ be a small performance impact though from making it strict. I only heard it offhand once, but apparently there's an extra is-evaluated check adding to all usages of the data constructor. Overall: I'd prefer to make it strict for correctness, and probably didn't have a good reason to avoid it in the first place.